In [24]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_log_error
from sklearn.metrics import mean_squared_error
import pandas as pd

In [25]:
df = pd.read_csv("training_dataset3.csv")

In [26]:
X = df[['Num_escuelas',
 'Num_hospitales',
 'Num_iglesias',
 'Num_zonas_tipicas',
  "Hora_llegada_1",                                          
  "Hora_llegada_2",
  "Hora_llegada_3",                                          
  "Hora_llegada_4",
  "Hora_llegada_5",                                          
  "Hora_llegada_6",
  "Hora_llegada_7",                                          
  "Hora_llegada_8",
  "Hora_llegada_9",                                          
  "Hora_llegada_10",
  "Hora_llegada_11",                                          
  "Hora_llegada_12",
  "Hora_llegada_13",                                          
  "Hora_llegada_14",
  "Hora_llegada_15",                                          
  "Hora_llegada_16",                                         
  "Hora_llegada_17",                                          
  "Hora_llegada_18",
  "Hora_llegada_19",                                          
  "Hora_llegada_20",
  "Hora_llegada_21",                                          
  "Hora_llegada_22",
  "Hora_llegada_23", ]]

y = df["Num_viajes"]

In [27]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.15, random_state=42)

In [28]:
linear_model = LinearRegression()
linear_model.fit(X, y)

LinearRegression()

In [29]:
from sklearn.metrics import mean_absolute_error as mae
from sklearn.metrics import mean_squared_log_error as msle
import numpy as np 

yfit_linear = pd.DataFrame(linear_model.predict(X_train), index = X_train.index, columns = ["Num_viajes"]).clip(0.)

ypred_linear = pd.DataFrame(linear_model.predict(X_test), index = X_test.index, columns = ["Num_viajes"]).clip(0.)


print('Linear RMSLE :', np.sqrt(msle(y_test, ypred_linear)))
print('Linear MAE :', mae(y_test, ypred_linear))


Linear RMSLE : 1.4166204730039076
Linear MAE : 293.0052847818318


Con las métricas iniciales, vemos claramente que el modelo representa una mejora sobre las iteraciones previas.

Ahora procederemos a evaluar más a fondo la precisión del modelo

# Evaluación del modelo

A continuación, calcularemos el porcentaje promedio de desviación entre la predicción y el valor real

In [30]:
dataframe = pd.DataFrame()

In [31]:
dataframe['Num_viajes'] = y_test

In [33]:
dataframe['Num_viajes_p'] = ypred_linear

In [34]:
dataframe['dif'] = abs(dataframe['Num_viajes'] - dataframe['Num_viajes_p'])

In [35]:
import math

In [36]:
dataframe['Num_viajes_p'] =  dataframe['Num_viajes_p'].apply(lambda x: math.ceil(x))

In [37]:
dataframe["Dif_perc"] =  dataframe["dif"] / dataframe["Num_viajes"] *100

In [38]:
dataframe["Dif_perc"].mean()

89.33346484561275

Vemos que tenemos una desviación media de 89%, es decir, en promedio esta será la diferencia entre los viajes que se predicen y los que ocurren en realidad